validators/valset/contract/fetcher.go

```diff
@@ -0,0 +1,76 @@
+package contract
+
+import (
+	"fmt"
+
+	"github.com/0xPolygon/polygon-edge/contracts/staking"
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/state"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/0xPolygon/polygon-edge/validators"
+)
+
+func FetchValidators(
+	validatorType validators.ValidatorType,
+	transition *state.Transition,
+	from types.Address,
+) (validators.Validators, error) {
+	switch validatorType {
+	case validators.ECDSAValidatorType:
+		return FetchECDSAValidators(transition, from)
+	case validators.BLSValidatorType:
+		return FetchBLSValidators(transition, from)
+	}
+
+	return nil, fmt.Errorf("unsupported validator type: %s", validatorType)
+}
+
+func FetchECDSAValidators(
+	transition *state.Transition,
+	from types.Address,
+) (*validators.ECDSAValidators, error) {
+	valAddrs, err := staking.QueryValidators(transition, from)
+	if err != nil {
+		return nil, err
+	}
+
+	ecdsaValidators := &validators.ECDSAValidators{}
+	for _, addr := range valAddrs {
+		if err := ecdsaValidators.Add(validators.NewECDSAValidator(addr)); err != nil {
+			return nil, err
+		}
+	}
+
+	return ecdsaValidators, nil
+}
+
+func FetchBLSValidators(
+	transition *state.Transition,
+	from types.Address,
+) (*validators.BLSValidators, error) {
+	valAddrs, err := staking.QueryValidators(transition, from)
+	if err != nil {
+		return nil, err
+	}
+
+	blsPublicKeys, err := staking.QueryBLSPublicKeys(transition, from)
+	if err != nil {
+		return nil, err
+	}
+
+	blsValidators := &validators.BLSValidators{}
+
+	for idx := range valAddrs {
+		// ignore the validator whose BLS Key is not set
```

Comment:
Why?

Reply:
BLS Validator must have address and BLS Public Key. In PoS, these data are stored in a contract. That's why need to skip a validator that didn't register BLS Public Key yet.

Reply:
Please add this as a comment  🙏

Reply:
@zivkovicmilos The code at latest commit has such comments